pinot-common/src/main/java/org/apache/pinot/common/utils/config/QueryWorkloadConfigUtils.java

```diff
@@ -211,24 +215,167 @@ public static List<String> validateQueryWorkloadConfig(QueryWorkloadConfig confi
+   *   <li>Checks for duplicate propagationEntity IDs</li>
+   *   <li>Validates cpuCostNs and memoryCostBytes for non-null/non-negative values</li>
+   *   <li>Ensures consistency in cost definitions across all entities (either all or none define costs)</li>
+   *   <li>Validates that total costs do not exceed provided limits (if any)</li>
+   *   <li>Validates any overrides within each entity for the same cost rules</li>
+   *   <li>Rewrites empty costs to evenly distribute parent limits if all entities have empty costs</li>
+   * </ul>
+   *
+   */
+  private static void validateEntityList(List<PropagationEntity> entities, String prefix,
+                                         List<String> errors, Long limitCpu, Long limitMem) {
+    if (entities == null || entities.isEmpty()) {
+      errors.add(prefix + " cannot be null or empty");
+      return;
+    }
+    Set<String> seenIds = new HashSet<>();
+    // Accumulate total CPU/memory costs to ensure they don't exceed enforcementProfile limits
+    long totalCpu = 0;
+    long totalMem = 0;
+    // Track whether costs are defined or empty to ensure consistency across all entities
+    int definedCount = 0;
+    int emptyCount = 0;
+    for (int i = 0; i < entities.size(); i++) {
+      PropagationEntity entity = entities.get(i);
+      String entityPrefix = prefix + "[" + i + "]";
+      if (entity == null) {
+        errors.add(entityPrefix + " cannot be null");
+        continue;
+      }
+      validateDuplicateEntity(entity.getEntity(), entityPrefix, seenIds, errors);
+      Long currentCpu = entity.getCpuCostNs();
+      Long currentMem = entity.getMemoryCostBytes();
+      // Both costs must be defined or both null
+      // If both are defined, add to totalCpu and totalMem for limit validation
+      // If both are null, do nothing
+      if (currentCpu != null && currentMem != null) {
+        totalCpu += costOrZero(entityPrefix, "cpuCostNs", currentCpu, errors);
+        totalMem += costOrZero(entityPrefix, "memoryCostBytes", currentMem, errors);
+        definedCount++;
+      } else if (currentCpu == null && currentMem == null) {
+        emptyCount++;
+      } else {
+        errors.add(entityPrefix + " must have both cpuCostNs and memoryCostBytes defined or both null");
+        break;
+      }
+      if (definedCount > 0 && emptyCount > 0) {
+        errors.add(prefix + " must have either all or none of the propagationEntities define costs");
+        break;
+      }
+      List<PropagationEntityOverrides> overrides = entity.getOverrides();
+      if (overrides != null && !overrides.isEmpty()) {
+        validateOverrides(overrides, entityPrefix, errors, currentCpu, currentMem);
+      }
+    }
+    validateLimits(totalCpu, totalMem, limitCpu, limitMem, prefix, errors);
+    // If no errors and all entities have empty costs, rewrite to evenly distribute enforcementProfile costs
+    if (errors.isEmpty() && definedCount == 0 && emptyCount == entities.size()) {
```

Comment:
errors will be non-empty is some entities have cost defined. So, can this check just be if(errors.isEmpty() && definedCount == 0)?